examples/rust/tcp-push-pop.rs

```diff
@@ -100,218 +100,209 @@ fn close(libos: &mut LibOS, sockqd: QDesc) {
     }
 }
 
-//======================================================================================================================
-// server()
-//======================================================================================================================
+// The TCP server.
+pub struct TcpServer {
+    /// Underlying libOS.
+    libos: LibOS,
+    /// Local socket queue descriptor.
+    sockqd: QDesc,
+    /// Accepted socket queue descriptor.
+    accepted_qd: Option<QDesc>,
+    /// The scatter-gather array.
+    sga: Option<demi_sgarray_t>,
+}
 
-fn server(local: SocketAddrV4) -> Result<()> {
-    let libos_name: LibOSName = match LibOSName::from_env() {
-        Ok(libos_name) => libos_name.into(),
-        Err(e) => anyhow::bail!("{:?}", e),
-    };
-    let mut libos: LibOS = match LibOS::new(libos_name) {
-        Ok(libos) => libos,
-        Err(e) => anyhow::bail!("failed to initialize libos: {:?}", e.cause),
-    };
-    let nbytes: usize = BUFFER_SIZE * 1024;
+// Implementation of the TCP server.
+impl TcpServer {
+    pub fn new(mut libos: LibOS) -> Result<Self> {
+        // Create the localsocket.
+        let sockqd: QDesc = match libos.socket(AF_INET, SOCK_STREAM, 0) {
+            Ok(sockqd) => sockqd,
+            Err(e) => anyhow::bail!("failed to create socket: {:?}", e),
+        };
 
-    // Setup peer.
-    let sockqd: QDesc = match libos.socket(AF_INET, SOCK_STREAM, 0) {
-        Ok(sockqd) => sockqd,
-        Err(e) => anyhow::bail!("failed to create socket: {:?}", e.cause),
-    };
-    match libos.bind(sockqd, local) {
-        Ok(()) => (),
-        Err(e) => {
-            // If error, free socket.
-            close(&mut libos, sockqd);
+        return Ok(Self {
+            libos,
+            sockqd,
+            accepted_qd: None,
+            sga: None,
+        });
+    }
+
+    pub fn run(&mut self, local: SocketAddrV4, fill_char: u8, buffer_size: usize) -> Result<()> {
+        let nbytes: usize = buffer_size * 1024;
+
+        if let Err(e) = self.libos.bind(self.sockqd, local) {
```

Comment:
Again, I think it would be fine to do this in the constructor. Then if the bind or listen fail, we don't also run drop()

Reply:
We discussed offline: we rely on the fact that the `drop()` function is not called if the constructor (`new()`) fails. So, if we `bind()` in `new()`, there's a chance that it fails and the socket may not be closed because `drop()` was not invoked.